{% extends "bootstrap/base.html" %}
{% import "bootstrap/fixes.html" as fixes %}
{% import "bootstrap/wtf.html" as wtf %}
{% block head %}{{super()}}{{fixes.ie8()}}{% endblock %}
{% block html_attribs %} lang="en"{% endblock %}
{% block metas %}
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
{% endblock %}

{% block title %}LISA{% endblock %}

{% block navbar %}
<nav class="navbar fixed-top navbar-expand-lg navbar-dark">
  <div class="container">
    <a href="/" class="navbar-brand">
      <img src="/static/lisa.jpg" height="55" width="40" alt="lisa">
      <span></span>LISA
    </a>
    <div id="navbar-nav-scroll">
      <ul class="navbar-nav bd-navbar-nav flex-row">
        <li class="nav-item active">
          <a class="nav-link nav-a" href="http://lisa.cistrome.org/">HOME</a>
        </li>

        <li class="nav-item active">
          <a class="nav-link nav-a" href="http://lisa.cistrome.org/new_gallery/new_gallery.html">Combined Human Gallery</a>
        </li>
        <li class="nav-item active">
          <a class="nav-link nav-a" href="http://lisa.cistrome.org/new_gallery_mm/new_gallery_mm.html">Combined Mouse Gallery</a>
        </li>

        <li class="nav-item active">
          <a class="nav-link nav-a" href="http://lisa.cistrome.org/doc">Documentation</a>
        </li>
      </ul>
    </div>
  </div>
</nav>
{% endblock %}

{% block styles %}
<!-- <link rel="stylesheet" href="http://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta/css/bootstrap.min.css" integrity="sha384-/Y6pD6FV/Vv2HJnA6t+vslU6fwYXjCFtcEpHbNJ0lyAFsXTsjBbfaDjzALeQsN6M" crossorigin="anonymous"> -->
<link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.1.1/css/bootstrap.min.css" integrity="sha384-WskhaSGFgHYWDcbwN70/dfYBj47jz9qbsMId/iRN3ewGhXQFZCSftd1LZCfmhktB" crossorigin="anonymous">
<link rel="stylesheet" href="{{ url_for('.static', filename='lisa.css') }}">
{% endblock %}

{% block content %}
<div class="row">
<div class="d-none d-xl-block col-xl-2 bd-toc">
<ul class="section-nav">
<li class="toc-entry toc-h2"><a href="#item-0">About</a></li>
<li class="toc-entry toc-h2"><a href="#item-1">Online version</a></li>
<ul>
<li class="toc-entry toc-h3"><a href="#item-1-1">Online usage</a></li>
<li class="toc-entry toc-h3"><a href="#item-1-2">Online output</a></li>
</ul>
</li>
<li class="toc-entry toc-h2"><a href="#item-2">Local command version</a>
<ul>
<li class="toc-entry toc-h3"><a href="#item-2-1">Local installation</a></li>
<li class="toc-entry toc-h3"><a href="#item-2-2">Local usage</a></li>
<li class="toc-entry toc-h3"><a href="#item-2-3">Local output</a></li>
</ul>
<li class="toc-entry toc-h2"><a href="#item-3">FAQ</a></li>
</li>
</ul>
</div>

<div class="col-sm-8 body3">
<div data-target="#navbar-example" data-offset="0">
  <h1 id="item-0">LISA</h2>

  <h4 id="item-0">About</h4>
  <p class="just">Transcriptional regulators (TRs), including transcription factors and chromatin regulators, play central role in gene regulation, and are essential in many biological processes. To rank TRs given geneset, we develop a computational biology framework, LISA, to evaluate and rank TRs effect by integrative modelling of chromatin landscapes and cistromes from Cistrome Data Browser (<a href="http://cistrome.org/db/">CistromeDB</a>). To run LISA online application or local command line, the only thing you need is the differential gene set derived from differential gene expression analysis of microarray and RNA-seq. Then you will be able to leverage the power of the most comprehensive TR ChIP-seq, DNase-seq, H3K27ac ChIP-seq database to discover the underlying regulatory TRs.</p>
  </p>
  </hr>

  <p class="just">LISA combines two models, epigenetics landscape in silico subtraction model and the cistrome regulatory potential model. The first model integrates epigenetics landscape with cistromes through in silico deletion strategy. First, the model will build a chromatin landscape model using the input gene set based on the comprehensive database of DNase-seq, or H3K27ac ChIP-seq. The model selects a subset of the cell type specific chromatin landscape. Then, the model will probe the effect of TRs on a gene in these landscapes by simulating deletions of chromatin signals on TR binding sites. These binding sites can come from the ChIP-seq experiment-derived data in CistromeDB, or the motif scan, or imputed TR binding sites from DREAM challenge. The second model uses only cistromes to model the TR regulatory potential on each gene. Finally, LISA would assign every binding site profile(ChIP-seq data, motif hit profile or imputed TF binding site) a score showing essentiality of the set of transcriptional regulator binding sites by comparing the effect between all the input genes and a fixed/user-customized background genes, and combined all the statistics from both models for each TR.</p>
  </hr>
</div>

  <h4 id="item-1">Website version</h4>
  <h5 id="item-1-1">Website Usage</h5>
  <p class="just">The input fields and options is shown in the input interface(Figure 1). User can input one or two gene set by using RefSeq ids or gene official symbols, usually the two gene sets can be obtained from up and down-regulated gene set from a differential gene expression analysis,and select the species the gene set originates from, for example, here are two gene sets for MYC knockdown in brain tumors: <a href="http://lisa.cistrome.org/gallery/83_up.txt">up-regulated gene set</a>, and <a href="http://lisa.cistrome.org/gallery/83_down.txt">down-regulated geneset</a>. If one gene set is given, it should be placed in "Gene Set 1", and the TF ranking list of that one gene list would be provided. If two are given, an additional figure of comparison between the two gene set analysis will be generated, a typical scenario would be input up and down regulated gene set from a set of expression profiles, the label for each gene set can be customized optionally. For gene set size, we would recommend 50 to 500 genes, since few genes will lower the predictive power of chromatin profile modelling. To build a model that can best represent the cell type specific chromatin landscape, the number of genes can not be too few. However, because in-silico deleting binding sites are time-consuming step for all the ChIP-seq profile and motif hit profile for all the input genes, the LISA running time would suffer from the number of genes when it goes beyond 500. There are DNase-seq, and H3K27ac ChIP-seq profiles for us to choose. From the set of transcription factor perturbation experiments used as benchmarks in the paper (see in <a href="http://lisa.cistrome.org/new_gallery/new_gallery.html">Human</a>/<a href="http://lisa.cistrome.org/new_gallery_mm/new_gallery_mm.html">Mouse</a> Gallery), we found that DNase-seq and H3K27ac are generally better than other chromatin profile, use can choose either to explore their gene set. User can select method based on TR ChIP-seq peak, based on in-silico knockout method, or both. A single run takes 5-10 minutes depending on the size of the gene set, user can optionally set an email to receive the result link.</p>
  </hr>

  <div class="figure1" id="id1" align="center">
  <img src="http://lisa.cistrome.org/static/Figure1.png" height="550" width="430">
  </div>
  <div align="center">Figure 1. LISA input interface</div>
  </hr>

  <h5 id="item-1-2">Website Output</h5>
  <p class="just">With two gene sets input, the MYC knockdown demo outputs <a href="http://lisa.cistrome.org/download/MYC_LISA_2019_04_30_0017080.114__hg38_result.html">the page</a>, which lists the following results as in Figure 2:</p>
  <div class="figure2" align="center">
  <img alt="_images/2.png" src="http://lisa.cistrome.org/static/Figure2.png" height="340" width="390">
  <div align="center">Figure 2. Results of LISA prediction of TR for MYC knockdown derived differentially expressed gene set</div>
  </div>

  <li>a pairwise scatterplot for comparison of the two gene set TF rankings based on LISA cis-elements knockout method based on ChIP-seq(Figure 3).</li>
  <div class="figure3" align="center">
  <img alt="_images/3.png" src="http://lisa.cistrome.org/static/Figure3.png" height="500" width="650">
  <div align="center">Figure 3. demo for LISA ranking of TF comparion between gene set 1 and gene set 2 by combined p-values</div>
  </div>

  <li>a pairwise scatterplot for comparison of the two gene set based on TF motif (Figure 4). </li>
  <div class="figure4" align="center">
  <img alt="_images/4.png" src="http://lisa.cistrome.org/static/Figure4.png" height="500" width="650">
  <div align="center">Figure 4. LISA TF ranking based on motif</div>
  </div>

  <p class="just">NOTE: These two figures can be saved to local disk, The x axis and y axis shows the -log10(p-value) of the TF ranking significance for either gene set. Once user mouse over the points, the corresponding TF ChIP-seq sample will be shown by Cistrome ID and TF name, user can see details information about the sample by searching CistromeDB with Cistrome ID.</p>

  <li>The TF ranking table of the LISA cis-elements knockout method based on ChIP-seq(Figure 4). </li>
  <li>The TF ranking table of the TF ranking from TF ChIP-seq peak(Figure 5). </li>
  <li>Selected chromatin profile and model performance. The annotation and the coefficient of the selected chromatin profile is selected (Figure 6). Along with the table, there is a ROCAUC curve to show the model performance.</li>

  <p class="just">NOTE: Click on each of the p-value for TF ranking or each sample for chromatin profile, the Cistrome dataset link will be shown below (Figure 5).</p>
  <li>The ranking table for Motif PWM based TF ranking (Figure 6).</li>
  <p class="just">NOTE: These tables above can be exported to csv or pdf to local disk. </p>

  <!--<li>WashU genome browser for gene set enrichment. We randomly picked arbitrarily 20 genes from the differential gene set to generate the genome tracks for the top ranked TF ChIP-seq, selected chromatin profile ChIP-seq for both gene set 1 (e.g. up regulated) and gene set 2 (e.g. down regulated) gene set (Figure 8).</li>-->
  <p class="just"></p>
  </hr>

  <div class="figure5" align="center">
  <img alt="_images/5.png" src="http://lisa.cistrome.org/static/Figure5.png" height="300" width="650">
  </div>
  <div align="center">Figure 5. LISA TF ranking based on TF-ChIP-seq peak method</div>

  <div class="figure6" align="center">
  <img alt="_images/6.png" src="http://lisa.cistrome.org/static/Figure6.png" height="310" width="650">
  </div>
  <div align="center">Figure 6. Selected chromatin profile, model performance, and demo for link to CistromeDB</div>

  <div class="figure7" align="center">
  <img alt="_images/7.png" src="http://lisa.cistrome.org/static/Figure7.png" height="400" width="650">
  </div>
  <div align="center">Figure 7. TF ranking based on chromatin profile knockout of motif hit site</div>

   <div class="figure1" align="center">
   <img alt="_images/8.png" src="http://cistrome.org/~qqin/8.png" height="400" width="650">
   </div>
   <!--
   <div align="center">Figure 8. gene set enrichment view mode on Wash U genome browser</div>
   -->

   <h4 id="item-2">Local command version</h4>
   <h4 id="item-2-1">Local installation</h4>
   <p class="just">LISA can be deployed as a command line version, which is installed using the following commands. </p>
   <figure class='highlight'>
   <pre>
    <code class="language-bash" data-lang="bash">
   ### Preparation of Anaconda environment
   wget -c https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh
   bash Miniconda3-latest-Linux-x86_64.sh
   export PATH="${HOME}/miniconda3/bin:$PATH"
   conda create -n lisa python=3.6 && conda config --add channels conda-forge && conda config --add channels bioconda
   conda activate lisa
   
   ### Installation
   conda install -c qinqian lisa
    </code>
   </pre>
   </figure>
   </hr>

   <h4 id="item-2-2">Local usage</h4>
    <p class="just">The LISA framework has three pipeline represented with three sub-parsers, <cite>model</cite>, <cite>multiple_fastq2bigwig</cite>, and <cite>multiple_bigwig2hdf</cite>;
    The TF rank method is run through the <cite>model</cite> sub-parser.</p>
    <h6>Run all the methods</h6>
    <td class="code"><div class="highlight"><pre>lisa model --method<span class="o">=</span><span class="s1">'all'</span> --web<span class="o">=</span>False <span class="se">\</span>
               --new_rp_h5<span class="o">=</span>None --new_count_h5<span class="o">=</span>None <span class="se">\</span>
               --species hg38 --epigenome <span class="s2">"['H3K27ac', 'DNase']"</span> <span class="se">\</span>
               --cluster<span class="o">=</span>True --covariates<span class="o">=</span>False <span class="se">\</span>
               --random<span class="o">=</span>True <span class="se">\</span>
               --background<span class="o">=</span>None <span class="se">\</span> ## option for user the specify background genes
               --prefix test_function1 <span class="se">\</span>
               --threads <span class="m">8</span> test.gene_symbol
    </pre>
    </td></tr></tbody></table>
    <p>Options:
    The <cite>test.gene_symbol</cite> is the gene set user input, which is a text file with one official gene symbol or refseq id per line, mixture of the both is not supported yet, this is used as a prefix for output, so put the file in a directory with written permission.</p>
    <p>-–epigenome option controls the histone mark, DNase-seq or both to use in the features.</p>
    <p>--background option can specify user provided text file containing one gene a row to compare with the target gene set, --background=None to use random selected 3000 genes.</p>
    <h6>Run only one method</h6></h3>
    <p>run only the knockout or beta (direct from TF ChIP-seq) method:
    --method option can control the method to run knockout or beta method, use it with the option –web=True to turn off the baseline method.</p>

    </td><td class="code"><div class="highlight"><pre>lisa model --method<span class="o">=</span><span class="s1">'knockout'</span> <span class="se">\</span>
               --web<span class="o">=</span>True --new_rp_h5<span class="o">=</span>None <span class="se">\</span>
               --new_count_h5<span class="o">=</span>None --species hg38 <span class="se">\</span>
               --epigenome <span class="s2">"['H3K27ac']"</span> <span class="se">\</span>
               --cluster<span class="o">=</span>False --covariates<span class="o">=</span>False <span class="se">\</span>
               --random<span class="o">=</span>True <span class="se">\</span>
               --background<span class="o">=</span>None <span class="se">\</span> ## option for user the specify background genes
               --prefix test_function1 --threads <span class="m">8</span> test.gene_symbol
    
    lisa model --method<span class="o">=</span><span class="s1">'beta'</span> <span class="se">\</span>
               --web<span class="o">=</span>True <span class="se">\</span>
               --new_rp_h5<span class="o">=</span>None --new_count_h5<span class="o">=</span>None <span class="se">\</span>
               --species hg38 --epigenome <span class="s2">"['H3K27ac']"</span> <span class="se">\</span>
               --cluster<span class="o">=</span>False --covariates<span class="o">=</span>False <span class="se">\</span>
               --random<span class="o">=</span>True <span class="se">\</span>
               --background<span class="o">=</span>None <span class="se">\</span> ## option for user the specify background genes
               --prefix test_function1 --threads <span class="m">8</span> test.gene_symbol
    </pre>
    </td></tr></tbody></table>
    
    <p>run only the cluster method to rank</p>
    </td><td class="code"><div class="highlight"><pre>lisa model --method<span class="o">=</span><span class="s1">'none'</span> <span class="se">\</span>
               --web<span class="o">=</span>True --new_rp_h5<span class="o">=</span>None <span class="se">\</span>
               --new_count_h5<span class="o">=</span>None --species hg38 <span class="se">\</span>
               --epigenome <span class="s2">"['H3K27ac']"</span> --cluster<span class="o">=</span>True <span class="se">\</span>
               --background<span class="o">=</span>None <span class="se">\</span> ## option for user the specify background genes
               --covariates<span class="o">=</span>False --random<span class="o">=</span>True <span class="se">\</span>
               --prefix test_function1 --threads <span class="m">8</span> test.gene_symbol
    </pre>
    </td></tr></tbody></table>

    <h6 id="item-2-3">Local Output</h6>
    <p>Take <em>AR.symbol</em> input and histone mark <em>H3K27ac</em> as an example. By using the above command, the key results are as follows,</p>
    <table border="0" class="table">
    <colgroup>
    <col width="26%">
    <col width="37%">
    <col width="36%">
    </colgroup>
    <thead valign="bottom">
    <tr class="row-odd"><th class="head">Output File</th>
    <th class="head">Content</th>
    <th class="head">Instruction</th>
    </tr>
    </thead>
    <tbody valign="top">
    <tr class="row-even"><td>AR.symbol.premodel_records.txt</td>
    <td>regulatory potential cluster-based regression performance and coefficients</td>
    <td>one row is PRAUC,AUC,lambda, 2nd row is coefficients</td>
    </tr>
    <tr class="row-odd"><td>AR.symbol_motif99_baseline.csv</td>
    <td>baseline method based on motif hit between differential &amp; control gene</td>
    <td>one row is motif_id|TF_name,p-value</td>
    </tr>
    <tr class="row-even"><td>AR.symbol_chipseq_baseline.csv</td>
    <td>baseline method based on chip-seq peak hit between differential &amp; control gene</td>
    <td>one row is CistromeID|TF_name,p-value</td>
    </tr>
    <tr class="row-odd"><td>AR.symbol.gene_symbol.H3K27ac.coefs.csv</td>
    <td>Regression coefficients for H3K27ac</td>
    <td>one row is CistromeID,coefficient,cell type, cell line, tissue</td>
    </tr>
    <tr class="row-even"><td>AR.symbol.H3K27ac.lisa_predicted_rp.csv</td>
    <td>predicted differential expression log probability</td>
    <td>one row is refseq,log probability</td>
    </tr>
    <tr class="row-odd"><td>AR.symbol.H3K27ac.gene_set.performance</td>
    <td>differential gene prediction</td>
    <td>one row is AUC,PRAUC</td>
    </tr>
    <tr class="row-even"><td>AR.symbol.lisa_direct.csv</td>
    <td>BETA TF ranks based on random background or epigenetic background</td>
    <td>one row is CistromeID|TF_name,p_value</td>
    </tr>
    <tr class="row-odd"><td>AR.symbol.gene_symbol.H3K27ac.motif99.p_value.csv</td>
    <td>motif insilco knockout result for H3K27ac</td>
    <td>one row is motif_id|TF_name,p-value</td>
    </tr>
    <tr class="row-even"><td>AR.symbol.H3K27ac.chipseq.p_value.csv</td>
    <td>TF ChIP-seq insilco knockout result for H3K27ac</td>
    <td>one row is motif_id|TF_name,p-value</td>
    </tr>
    <tr class="row-odd"><td>AR.symbol.gene_symbol.H3K27ac.motif99.entropy_rank.csv</td>
    <td>motif-based TF ranks from KL divergence method</td>
    <td>one row is  motif_id|TF_name,KLD,rank</td>
    </tr>
    <tr class="row-even"><td>AR.symbol.gene_symbol.H3K27ac.chipseq.entropy_rank.csv</td>
    <td>ChIP-seq-based TF ranks from KL divergence method</td>
    <td>one row is motif_id|TF_name,KLD,rank|</td>
    </tr>
    </tbody>
    </table>

    <h4 id="item-3">FAQ</h4>
    <h6>How many datasets are there?</h6>
    <p>The 1st version of CistromeDB is incorporated, the dataset statistics is as the table below, NOTE: the data used for current LISA online is updated to CistromeDB (201801). </p>
    <div align="center">Table 1. Chromatin profile data and TF/CR ChIP-seq dataset</div>
    <table border="0" class="table">
    <colgroup>
    <col width="34%">
    <col width="34%">
    <col width="31%">
    </colgroup>
    <thead valign="bottom">
    <tr class="row-odd"><th class="head">Category</th>
    <th class="head">Human</th>
    <th class="head">Mouse</th>
    </tr>
    </thead>
    <tbody valign="top">
    <tr class="row-even"><td>DNase-seq</td>
    <td>1,110</td>
    <td>359</td>
    </tr>
    <tr class="row-odd"><td>H3K27ac ChIP-seq</td>
    <td>1,545</td>
    <td>1,031</td>
    </tr>
    <!--
    <tr class="row-even"><td>H3K27me3</td>
    <td>842</td>
    <td>932</td>
    </tr>
    <tr class="row-odd"><td>H3K4me3</td>
    <td>1,432</td>
    <td>1,032</td>
    </tr>
    <tr class="row-even"><td>H3K4me1</td>
    <td>726</td>
    <td>590</td>
    </tr>
    -->
    <tr class="row-odd"><td>TF and CR ChIP-seq</td>
    <td>6,773</td>
    <td>4,363</td>
    </tr>
    <!--
    <tr class="row-even"><td>Total</td>
    <td>12,428</td>
    <td>8,307</td>
    </tr>
    -->
    </tbody>
    </table>

    <hr>
  </div>
</div>

<nav class="navbar fixed-bottom navbar-dark">
  <div class="container">
    <a href="/" class="navbar-brand">
      <img src="http://liulab.dfci.harvard.edu/WEBSITE/images/Global/logofinal.png" height="46" width="58">
    </a>
    <ul class="navbar-nav bd-navbar-nav flex-row">
    <li class="nav-item active">
     <span>Copyright @2018 Liu lab</span>
    </li>
    </ul>
  </div>
</nav>
{% endblock %}


{% block scripts %}
<!-- {{ super() }} -->
<!-- <script src="https://code.jquery.com/jquery-3.2.1.min.js"></script> -->
<!-- <script src="{{ url_for('.static', filename='jquery-3.2.1.min.js') }}"></script> -->
<!-- <script src="{{ url_for('.static', filename='popper.min.js') }}"></script> -->
<script src="https://code.jquery.com/jquery-3.3.1.slim.min.js" integrity="sha384-q8i/X+965DzO0rT7abK41JStQIAqVgRVzpbzo5smXKp4YfRvH+8abtTE1Pi6jizo" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.1/js/bootstrap.min.js" integrity="sha384-smHYKdLADwkXOn1EmN1qk/HfnUcbVRZyYmZ4qpPea6sjB/pTJ0euyQp0Mk8ck+5T" crossorigin="anonymous"></script>

<!-- <script src="http://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta/js/bootstrap.min.js" integrity="sha384-h0AbiXch4ZDo7tp9hKZ4TsHbi047NrKGLO3SEJAg45jXxnGIfYzk4Si90RDIqNm1" crossorigin="anonymous"></script> -->
<!-- <script src="http://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.3/jquery.min.js"></script> -->
<script src="{{ url_for('.static', filename='lisa.js') }}" type="text/javascript"></script>
<!--<script>
$('#spyOnThis').scrollspy();
</script>
-->
{% endblock %}
